Battery critical/low actions

WickedShell opened the pull request:
This builds upon #7177 to enable different failsafe levels per connected battery, as well as allowing the vehicle to define a critical and a low battery action. This also serves as a replacement for #5000.

This is a WIP at the moment as this has only touched plane so far. Given the scope of the changes, and how it touches all the vehicles I'd like to get feedback on the proposed approach before I look at reworking copter/rover/sub to all support this as well. (I don't really know what the best actions are for a lot of those vehicles, although I think that some like terminate can be supported on all vehicles if AFS is enabled. (The initial pr would probably just present the options to have the current behavior on those vehicles unless someone wants to steer further).

A couple of points worth noting here at a top leve that I would appreciate feedback onl:
1. As I've presented it right now both critical and low battery actions use the same timeout duration to trip based on battery voltage, (they are both referencing the `BATT_LOW_TIMER` parameter). I can't see a strong reason why you would want a different timeout for critical vs long, and would prefer to rename this as `BATT_FS_TIMER` but it would be yet another param rename. Having 2 timers is easy, but I'm not sure its worth the param space.
2. I left the FS actions in the vehicle library to determine what to do on low battery. If we had the documentation changes PR that allows us to tag a param doc as vehicle specific then we could move this parameter into the AP_BattMonitor level.

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -0,0 +1,56 @@
+    AP_Float _curr_amp_per_volt;       /// voltage on current pin multiplied by this to calculate current in amps
+    AP_Float _curr_amp_offset;         /// offset voltage that is subtracted from current pin before conversion to amps
+    AP_Int32 _pack_capacity;           /// battery pack capacity less reserve in mAh
+    AP_Int16 _watt_max;                /// max battery power allowed. Reduce max throttle to reduce current to satisfy t    his limit
```
whitespaces in comment

amilcarlucas commented:
https://github.com/ArduPilot/ardupilot/pull/7432 provides a way to customize what happens when a FS gets triggered. So these two can get combined ?

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -51,18 +61,19 @@ class AP_BattMonitor
-        float       voltage_resting_estimate; // voltage with sag removed based on current and resistance estimate
-        float       resistance;         // resistance calculated by comparing resting voltage vs in flight voltage
-        bool        healthy;            // battery monitor is communicating correctly
+        cells       cell_voltages;             // battery cell voltages in millivolts, 10 cells matches the MAVLink spec
```
Does the added value of indentation surpass the value of breaking git blame ?

WickedShell replied:
I've been caught! :)

WickedShell commented:
@amilcarlucas I'd hold off on reviewing to much yet, been pushing more for linking to others :) Moving a large pile of additional stuff into AP_BattMonitor first.

WickedShell commented:
This has been pushed with support for all vehicles (Tracker is a no op at the moment, but the skeleton is there for others).

Summary of changes:
- Compass current compensation is automatically polled from the battery if needed, and is now consistent across all vehicles
- All vehicles skip running fail safes while disarmed based on soft armed. (Sub and Copter were basing it on `motors->armed()`)
- Copter special case for USB disabling failsafes has been removed
- Copter will now trigger failsafes during motor test and compass motor cal. This behavior can be restored to what it was, but the rational for running failsafes during this is to let someone know if they are risking damaging a battery, and in the cal step running at an out of range battery voltage seems like it could produce different results. This needs a copter expert to review (particularly to make sure the copter can't accidentally takeoff if it failsafes during a motor test).
- All vehicles now have a low and critical threshold set per battery, with a selectable set of actions for both low and critical (again per battery). This allows you to select a different failsafe action for your propulsion battery and avionics battery.
- Sub no longer broadcasts a statustext every 20 seconds about low battery
- Sub now has a param conversion table for the first time
- Rover and tracker both report the battery in SYS_STATUS the same way copter/plane do. (Always present, enabled when there is an instance loaded, unhealthy if a failsafe occurs).

Outstanding Issues:
- Needs a form of failsafe priorities. At the moment it assumes that each increasing failsafe action parameter value is of higher importance then the previous. While this works well on plane at the moment it doesn't scale if we insert something of middle priority, and it doesn't work for Copter/Sub at all at the moment. The two solutions for this are to either break the param conversion code (which while I think it's valid it's probably not popular), or to pass a table that maps failsafe actions to priority levels. (This is probably the best bet, but I wanted to source other opinions before I did it).

Needed Testing (SITL is fine):
- [ ] Copter
- [ ] Plane (I've partially finished this, still need to validate the param conversion/update code, as well as the interaction with a second monitor enabled)
- [ ] Sub
- [ ] Rover

If the vehicle maintainers can chime in at this point that would be great!

amilcarlucas commented:
Great work on unifying vehicle behavior and moving code from vehicle to the libraries :)

WickedShell commented on ArduCopter/Parameters.cpp:
```diff
@@ -148,29 +148,6 @@ const AP_Param::Info Copter::var_info[] = {
-    // @Units: V
-    // @Increment: 0.1
-    // @User: Standard
-    GSCALAR(fs_batt_voltage,        "FS_BATT_VOLTAGE", FS_BATT_VOLTAGE_DEFAULT),
```
@rmackay9 is it really desired to provide a default voltage here? I can move that into batt monitor in a special case for copter, but it does seem questionable to me given that it's wrong far more often then it's right.

amilcarlucas commented:
It would be nice if the description text of this PR gets added to at least one of the commits. Maybe the AP_BattMonitor one.

khancyr commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -115,9 +127,15 @@ class AP_BattMonitor
+    void check_failsafes(void); // checks all batteries failsafes
+
+    /// returns true if a battery failsafe has ever been triggered
+    bool has_failsafed(void) { return _has_triggered_failsafe; };
```
const

khancyr commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -115,9 +127,15 @@ class AP_BattMonitor
+    bool has_failsafed(void) { return _has_triggered_failsafe; };
+
+    /// returns the highest failsafe action that has been triggered
+    int8_t get_highest_failsafe_action(void) { return _highest_failsafe_action; };
```
const

khancyr commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -287,14 +291,56 @@ int32_t AP_BattMonitor::pack_capacity_mah(uint8_t instance) const
+{
+    if (hal.util->get_soft_armed()) {
+        for (uint8_t i = 0; i < _num_instances; i++) {
+            BatteryFailsafe type = check_failsafe(i);
```
const

khancyr commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -287,14 +291,56 @@ int32_t AP_BattMonitor::pack_capacity_mah(uint8_t instance) const
+        return BatteryFailsafe_None;
     }
 
+    uint32_t now = AP_HAL::millis();
```
const

khancyr commented on libraries/AP_Compass/AP_Compass.cpp:
```diff
@@ -1187,7 +1186,7 @@ void Compass::motor_compensation_type(const uint8_t comp_type)
     if (_motor_comp_type <= AP_COMPASS_MOT_COMP_CURRENT && _motor_comp_type != (int8_t)comp_type) {
         _motor_comp_type = (int8_t)comp_type;
-        _thr_or_curr = 0;                               // set current current or throttle to zero
+        _thr = 0; // set current current or throttle to zero
```
change comment !

khancyr commented:
Looks good at first read, the FS_Action could be move to AP_Batterie as we now have the parser to generate param doc according to each vehicle.
I will test that as soon as I can

WickedShell commented:
> the FS_Action could be move to AP_Batterie
@khancyr I'm not following sorry. The action per vehicle/level did move to AP_BattMonitor, which param are you referring to?

WickedShell commented:
- Added priority lists for failsafes on the vehicles
- Fixed Copter autotest
- Added a lazy set second of voltage/current pins to the simulator intended for battery useage, they are just the normal battery pins scaled by 25% (I just needed something to test the interactions of two batteries in SITL)

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
 
+    // @Param: LOW_VOLT
+    // @DisplayName: Battery low voltage
+    // @Description: Battery voltage to trigger a low battery failsafe. Set to 0 to disable. If the battery voltage drops below this voltage continuously for 10 seconds then the vehicle will switch to RTL mode.
```
Maybe "this voltage continuously for 10 seconds" to "this voltage for more than 10 seconds "

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+    AP_GROUPINFO("LOW_VOLT", 12, AP_BattMonitor_Params, _low_voltage, DEFAULT_LOW_BATTERY_VOLTAGE),
+
+    // @Param: LOW_MAH
+    // @DisplayName: Battery low capacity
```
I'd be tempted by "Low battery capacity" or "Low battery capacity failsafe trigger level".

WickedShell replied:
I like the former more. Updated, I'm on the fence if the voltage should be arranged similarly. Does any GCS use this display name to sort a list? (IE does this change risk breaking parameter grouping for them?)

WickedShell replied:
Just checked with @meee1 it's grouped by the param name rather then the display name, so keeping the more natural version is fine. Should the voltage one be rearranged as well then? (IE `Low battery voltage`)

hamishwillee replied:
Up to you, but IMO they should be consistent.

WickedShell replied:
Done.

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+
+    // @Param: LOW_MAH
+    // @DisplayName: Battery low capacity
+    // @Description: Battery capacity remaining to trigger a low battery failsafe. Set to 0 to disable battery remaining failsafe. If the battery capacity drops below this level the vehicle will switch to RTL mode.
```
Battery capacity at which the low battery failsafe is triggered.

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+
+    // @Param: CRT_VOLT
+    // @DisplayName: Battery critical voltage
+    // @Description: Battery voltage to trigger a criticial battery failsafe. Set to 0 to disable. If the battery voltage drops below this voltage continuously for 10 seconds then the vehicle will switch to RTL mode.
```
"that triggers" rather than "to trigger" (but this is OK if you can't be bothered changing - no one will misinterpret).

hamishwillee replied:
"battery voltage drops below this voltage continuously for 10 seconds " -> battery voltage drops below this voltage for more than 10 seconds

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+
+    // @Param: CRT_MAH
+    // @DisplayName: Battery critical capacity
+    // @Description: Battery capacity remaining to trigger critical failsafe. Set to 0 to disable the critical battery remaining failsafe. If the battery capacity drops below this level the vehicle will switch to RTL mode.
```
Battery capacity level below which the battery-critical failsafe is triggered. Set to 0 to disable the the failsafe.

(there is something about "Battery capacity remaining" that annoys me. Maybe "Remaining battery capacity" would work better.)

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+
+    // @Param: FS_LOW_ACT
+    // @DisplayName: Low battery failsafe action
+    // @Description: What action the vehicle should preform if it hits a low battery failsafe
```
The action that the vehicle should perform ...

(note too perform, not preform)

hamishwillee commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+
+    // @Param: FS_CRT_ACT
+    // @DisplayName: Critical battery failsafe action
+    // @Description: What action the vehicle should preform if it hits a critical battery failsafe
```
The action that the vehicle should perform ...

jaxxzer commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -295,14 +302,68 @@ int32_t AP_BattMonitor::pack_capacity_mah(uint8_t instance) const
 
- /// exhausted - returns true if the voltage remains below the low_voltage for 10 seconds or remaining capacity falls below min_capacity_mah
-bool AP_BattMonitor::exhausted(uint8_t instance, float low_voltage, float min_capacity_mah)
+void AP_BattMonitor::check_failsafes(void)
```
This looks ok, but I think anyone else who visits this bit of code will appreciate a comment with a high-level rundown of what is going on here.

jaxxzer commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -83,6 +89,57 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
+
+    // @Param: CRT_VOLT
+    // @DisplayName: Battery critical voltage
+    // @Description: Battery voltage to trigger a criticial battery failsafe. Set to 0 to disable. If the battery voltage drops below this voltage continuously for 10 seconds then the vehicle will switch to RTL mode.
```
These descriptions need to be updated to replace 'RTL mode' with the current implementation.

jaxxzer commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -295,14 +302,68 @@ int32_t AP_BattMonitor::pack_capacity_mah(uint8_t instance) const
+            _has_triggered_failsafe = true;
+            AP_Notify::flags.failsafe_battery = true;
+            state[i].failsafe = type;
+
```
Should we/do we need to assert that the critical action setting have >= priority than the low action setting?

WickedShell replied:
I think it's a good/worthy idea, but I think it's something that belongs in an arming check, and I'd like to address that in a different/later PR rather then keep creeping the scope on this one. (In general I want to tweak the battery arming checks a bit, so I'd like to just do that in one pass)

WickedShell commented on ArduSub/Sub.h:
```diff
@@ -266,26 +266,24 @@ class Sub : public AP_HAL::HAL::Callbacks {
         uint32_t terrain_last_failure_ms;   // the most recent time terrain data access failed
-        uint32_t last_battery_warn_ms; // last time a battery failsafe warning was sent to gcs
         uint32_t last_crash_warn_ms; // last time a crash warning was sent to gcs
         uint32_t last_ekf_warn_ms; // last time an ekf warning was sent to gcs
```
Alright so this is interesting. The first one is when I was removing the battery tag as it needed to be updated. This in turn brought up the fact that all the comments about which bit it was wrong, so I fixed it. Once I did that though rearranging the structure (which didn't change the diff size appreciably), actually packs into a smaller structure and uses less memory.

---

TL;DR rearranged structure to reduce memory usage and fix the comments.

WickedShell commented:
@hamishwillee Good catches, thanks. I pushed up my changes. My only concern is that the same doc string will be emitted for both battery 1 and battery 2. What this means is that the in GCS docs (or online) would say on BATT2_LOW_VOLT to look at BATT_FS_LOW_ACT which is incorrect. I tried striking the middleground of just saying `FS_LOW_ACT` as that's appropriately generic but without support for the feature I proposed in #7908 I'm not sure that this is better/less confusing. Maybe BATT(2)_FS_LOW_ACT? But even that is probably surprising to users.

hamishwillee commented:
> My only concern is that the same doc string will be emitted for both battery 1 and battery 2. What this means is that the in GCS docs (or online) would say on BATT2_LOW_VOLT to look at BATT_FS_LOW_ACT which is incorrect. I tried striking the middleground of just saying FS_LOW_ACT as that's appropriately generic but without support for the feature I proposed in #7908 I'm not sure that this is better/less confusing. Maybe BATT(2)_FS_LOW_ACT? But even that is probably surprising to users.


@WickedShell OK, so I think you are saying that normally you can refer to a battery generically, but when you are referring to a parameter you want to/have to refer to the specific battery variant (e.g. BATT2_) or the docs will be wrong? 

Yes, sounds like a problem. Unfortunately I don't know how this works. Are we SURE it is a problem and there isn't a magic solution already in place? I only ask because if you look at: [BATT2_VOLT_MULT: Voltage Multiplier](http://ardupilot.org/copter/docs/parameters.html#batt2-volt-mult-voltage-multiplier) the docs say `Used to convert the voltage of the voltage sensing pin (BATT2_VOLT_PIN)`, but the [generic doc](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp#L28) says `BATT2_VOLT_PIN`. So this has been solved "somehow".

WickedShell commented:
@hamishwillee Unfortunately that was correct before because the parameters for BATT2 were indvidiually listed/documented, now that per battery instances are all groups themselves (which lets us easily scale up the number of instances/not need to cut/paste/update so many lines), the docs are wrong. (The wiki docs have not been updated since Dec 15th 2017 which is a separate issue). If you generate the docs these days on what's in master this is BATT2_MULT:

---
## BATT2_VOLT_MULT: Voltage Multiplier

*Note: This parameter is for advanced users*

Used to convert the voltage of the voltage sensing pin (BATT_VOLT_PIN) to the actual battery's voltage (pin_voltage * VOLT_MULT). For the 3DR Power brick on APM2 or Pixhawk, this should be set to 10.1. For the Pixhawk with the 3DR 4in1 ESC this should be 12.02. For the PX using the PX4IO power supply this should be set to 1.
---

hamishwillee commented:
> @hamishwillee Unfortunately that was correct before because the parameters for BATT2 were individually listed/documented, now that per battery instances are all groups themselves (which lets us easily scale up ....
Options. 
1. Use BATTx_VOLT_PIN or similar to refer to other parameters
2. Tooling does search/replace for any text starting with format BATT_ with current BATTX (e.g. BATT2_)
3. Support additional markup like `voltage sensing pin ({CUR}_VOLT_PIN)` to tell the system the specific items that are current_version

I think that is pretty much what you suggested and what you can do. IMO I would do 1 for now, and perhaps look at 3 later?

WickedShell commented:
Yup. BATTx_ is probably clearer then the parenthesis thing I was going for so I went with that for now. Option 2 is definitely the best long term.

hamishwillee commented:
> Yup. BATTx_ is probably clearer then the parenthesis thing I was going for so I went with that for now. Option 2 is definitely the best long term.

Probably Option 3 because there are cases where BATT_ represents a parameter for all batteries - e.g. a failsafe behaviour, and you wouldn't want that automatically turned into BATTx_ automatically.

WickedShell commented:
Yeah that's why my proposal was to have a tag that tells the script when to replace it with the prefix or not. (So it's then opt in on a per param doc basis)

hamishwillee commented:
> Yeah that's why my proposal was to have a tag that tells the script when to replace it with the prefix or not. (So it's then opt in on a per param doc basis)

Indeed! After some thought, that if we use option 1, then x can act as the marker - combining 2 and 3.

So our source look like this:
> Used to convert the voltage of the voltage sensing pin (BATTx_VOLT_PIN) to the actual battery's voltage (pin_voltage * VOLT_MULT). 

Which works now. But our tooling looks for pattern BATTx_ and replaces/removes the x as appropriate. I think generically the string CURRENTGROUPx_ (with letter case as shown) is probably safe to use for a generic search replace.

WickedShell commented:
@hamishwillee The good news is Peter was feeling eager and tackled the issue in #7911. So you can now do `@PREFIX@VOLT_MULT` and this will generate documentation for both `BATT_VOLT_MULT` and `BATT2_VOLT_MULT` depending on which param group this is. (I've pushed up the changes here that use that)

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -53,18 +63,20 @@ class AP_BattMonitor
-        float       resistance;         // resistance calculated by comparing resting voltage vs in flight voltage
-        bool        healthy;            // battery monitor is communicating correctly
+        cells       cell_voltages;             // battery cell voltages in millivolts, 10 cells matches the MAVLink spec
+        float       voltage;                   // voltage in volts
```
If you are going to break git history here, please replace `volts` with `Volt`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -53,18 +63,20 @@ class AP_BattMonitor
-        bool        healthy;            // battery monitor is communicating correctly
+        cells       cell_voltages;             // battery cell voltages in millivolts, 10 cells matches the MAVLink spec
+        float       voltage;                   // voltage in volts
+        float       current_amps;              // current in amperes
```
If you are going to break git history here, please replace `amperes` with `Ampere`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -53,18 +63,20 @@ class AP_BattMonitor
+        cells       cell_voltages;             // battery cell voltages in millivolts, 10 cells matches the MAVLink spec
+        float       voltage;                   // voltage in volts
+        float       current_amps;              // current in amperes
+        float       consumed_mah;              // total current draw in milliamp hours since start-up
```
If you are going to break git history here, please replace `milliamp hours` with `milliampere hours`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -53,18 +63,20 @@ class AP_BattMonitor
+        float       voltage;                   // voltage in volts
+        float       current_amps;              // current in amperes
+        float       consumed_mah;              // total current draw in milliamp hours since start-up
+        float       consumed_wh;               // total energy consumed in Wh since start-up
```
If you are going to break git history here, please replace `Wh` with `Watt hours`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -53,18 +63,20 @@ class AP_BattMonitor
+        uint32_t    last_time_micros;          // time when voltage and current was last read
+        uint32_t    low_voltage_start_ms;      // time when voltage dropped below the minimum
+        uint32_t    critical_voltage_start_ms; // critical voltage failsafe start timer in milliseconds
+        float       temperature;               // battery temperature in celsius
```
If you are going to break git history here, please replace `celsius` with `degrees celsius`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor.h:
```diff
@@ -53,18 +63,20 @@ class AP_BattMonitor
+        float       temperature;               // battery temperature in celsius
+        uint32_t    temperature_time;          // timestamp of the last received temperature message
+        float       voltage_resting_estimate;  // voltage with sag removed based on current and resistance estimate
+        float       resistance;                // resistance calculated by comparing resting voltage vs in flight voltage
```
Add `in Ohm`

khancyr commented:
Working on skyviper ! I test report on batt_low, and terminate on batt_crt, working as expected

rmackay9 commented:
I guess having the battery failsafe right inside the battery sensor library is the best thing to do?  It could be separated out similar to how the compass calibration class is separated.

rmackay9 commented:
I'm happy for this to go in.  I have this slight concern about moving the failsafe actions down into the battery library but I don't really have a better suggestion and overall it's a great improvement.

OXINARF commented on APMrover2/sensors.cpp:
```diff
@@ -312,6 +312,10 @@ void Rover::update_sensor_status_flags(void)
     }
 
+    if (battery.num_instances() > 0) {
+        control_sensors_enabled |= MAV_SYS_STATUS_SENSOR_BATTERY;
```
You forgot to remove it from the default enabled sensors above.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -295,14 +302,68 @@ int32_t AP_BattMonitor::pack_capacity_mah(uint8_t instance) const
+            const char *type_str = nullptr;
+            switch (type) {
+                case AP_BattMonitor::BatteryFailsafe_None:
+                    return; // should not have been called in this case
```
Very good catch.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -315,28 +376,47 @@ bool AP_BattMonitor::exhausted(uint8_t instance, float low_voltage, float min_ca
+        // this is the first time our voltage has dropped below minimum so start timer
+        if (state[instance].critical_voltage_start_ms == 0) {
+            state[instance].critical_voltage_start_ms = now;
+        } else if (_params[instance]._low_voltage_timeout > 0 &&
```
Per the docs `A value of zero disables low voltage errors.` I didn't write it/support the original behavior but that is the expected legacy behavior at this point.

WickedShell commented on ArduPlane/sensors.cpp:
```diff
@@ -219,7 +206,7 @@ void Plane::update_sensor_status_flags(void)
     }
 
-    if (g.fs_batt_voltage > 0 || g.fs_batt_mah > 0) {
+    if (battery.num_instances() > 0) {
```
This was updated across all vehicles to be consistent. Always present, enabled when num_instances() > 0, healthy as long as the lib is healthy and not in failsafe.

WickedShell commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -1021,3 +1021,5 @@ float Plane::tecs_hgt_afe(void)
 
 AP_HAL_MAIN_CALLBACKS(&plane);
+
+constexpr int8_t Plane::_failsafe_priorities[5];
```
Located below the scheduler in each vehicle.cpp

WickedShell commented on ArduCopter/compassmot.cpp:
```diff
@@ -110,7 +110,6 @@ MAV_RESULT Copter::mavlink_compassmot(mavlink_channel_t chan)
 
     // disable throttle and battery failsafe
     g.failsafe_throttle = FS_THR_DISABLED;
-    g.failsafe_battery_enabled = FS_BATT_DISABLED;
```
No, and this is something that really needs to be tested. I don't have a copter and am not familiar enough to check that the vehicle might do something uprising here. We shouldn't be testing if we are at failsafe triggering levels, but we absolutely don't want to do something surprising (like RTL) here.

WickedShell commented on ArduCopter/events.cpp:
```diff
@@ -16,8 +16,10 @@ void Copter::failsafe_radio_on_event()
-            // continue landing
+        } else if (control_mode == LAND &&
+                   battery.has_failsafed() &&
+                   battery.get_highest_failsafe_priority() >= FAILSAFE_LAND_PRIORITY) {
```
Good catch. The removal of control mode is a behavior change, and probably not a desired one. Consider a low battery failsafe that starts a landing, but the user takes over in stabilize for example to avoid hitting something. That's perfectly reasonable, but on a radio failsafe we shouldn't stay in stabilize for example. (If we wanted to retrigger land then I can see that being very reasonable, although that is a behavior change I don't want to push at the moment)

OXINARF replied:
Ok, I guess that's fine. This is going to be the next issue we'll have: priorities between different failsafes.

WickedShell commented on ArduCopter/events.cpp:
```diff
@@ -46,37 +48,39 @@ void Copter::failsafe_radio_off_event()
+    } else {
+        switch ((Failsafe_Action)action) {
+            case Failsafe_Action_None:
+                return; // this is unreachable, but include it to keep the compiler happy
```
At one point I was special casing the none action and wasn't even handing it to the vehicles, but for flexibility reasons I decided we should still call the vehicle, the comment is stale.

WickedShell commented on ArduCopter/motor_test.cpp:
```diff
@@ -155,7 +155,6 @@ MAV_RESULT Copter::mavlink_motor_test_start(mavlink_channel_t chan, uint8_t moto
 
             // disable throttle, battery and gps failsafe
             g.failsafe_throttle = FS_THR_DISABLED;
-            g.failsafe_battery_enabled = FS_BATT_DISABLED;
```
Has not, same comment as above.

WickedShell commented on ArduCopter/AP_Arming.cpp:
```diff
@@ -162,14 +160,6 @@ bool AP_Arming_Copter::board_voltage_checks(bool display_failure)
             return true;
```
Yeah. The intention is to try and back off from this special case of the USB stuff. If we have to reintroduce it fine, but it should be a common behaviour in that case.

WickedShell commented on ArduSub/sensors.cpp:
```diff
@@ -172,13 +172,6 @@ void Sub::read_battery(void)
     }
-
-    if (battery.has_current()) {
-        motors.set_current(battery.current_amps());
```
This was a mistake. I've now expanded this further made AP_Motors use the battery singleton, which actually fixed a divide by 0 as well.

WickedShell commented on APMrover2/Rover.h:
```diff
@@ -553,6 +556,28 @@ class Rover : public AP_HAL::HAL::Callbacks {
+        Failsafe_Action_Hold          = 2,
+        Failsafe_Action_SmartRTL      = 3,
+        Failsafe_Action_SmartRTL_Hold = 4,
+        Failsafe_Action_Terminate     = 5
```
This actually matches the order in `APMrover2/failsafe.cpp` for the g.fs_action parameter. The next logical step is to make the normal failsafe handler use this same enum (and actually collapse the handling further). (This future intent is actually why these are called Failsafe_Action instead of Battery_Failsafe)

WickedShell commented:
I believe I've addressed all of @OXINARF's comments. Along the way AP_Motors also now uses the battery singleton, and catches a possible divide by 0.

WickedShell commented:
Argh, in the ever increasing scope creep that is this PR I'm pretty sure I just enabled the battery motor compensation code on quadplane. Will need to touch base with @tridge to confirm there aren't any  implications of this.

OXINARF commented on ArduSub/GCS_Mavlink.cpp:
```diff
@@ -1657,4 +1665,4 @@ void GCS_MAVLINK_Sub::set_ekf_origin(const Location& loc)
 
 // dummy method to avoid linking AFS
-bool AP_AdvancedFailsafe::gcs_terminate(bool should_terminate) { return false; }
+bool AP_AdvancedFailsafe::gcs_terminate(bool should_terminate, const char *src) { return false; }
```
src -> reason

OXINARF commented on AntennaTracker/GCS_Mavlink.cpp:
```diff
@@ -830,4 +830,4 @@ void AP_Camera::send_feedback(mavlink_channel_t chan) {}
 
 // dummy method to avoid linking AFS
-bool AP_AdvancedFailsafe::gcs_terminate(bool should_terminate) {return false;}
+bool AP_AdvancedFailsafe::gcs_terminate(bool should_terminate, const char *src) {return false;}
```
src -> reason

OXINARF commented on Tools/Replay/Replay.cpp:
```diff
@@ -954,6 +954,6 @@ GCS_Dummy _gcs;
 // dummy methods to avoid linking with these libraries
 void AP_Camera::send_feedback(mavlink_channel_t) {}
-bool AP_AdvancedFailsafe::gcs_terminate(bool should_terminate) { return false; }
+bool AP_AdvancedFailsafe::gcs_terminate(bool should_terminate, const char *src) { return false; }
```
src -> reason

OXINARF commented on libraries/AP_HAL_SITL/SITL_State.h:
```diff
@@ -62,6 +62,8 @@ class HALSITL::SITL_State {
     uint16_t voltage_pin_value;  // pin 13
     uint16_t current_pin_value;  // pin 12
+    uint16_t voltage2_pin_value;  // pin 14
+    uint16_t current2_pin_value;  // pin 15
```
Still wrong, voltage is pin 15, current pin 14 🙂

OXINARF commented:
So besides the small nitpicking above, what is missing?

- check motor test and motor compass compensation don't have issues when a battery failsafe happens
- re-add the pre-arm check for battery voltage before ever arming, which we've agreed you'll do in another PR soon

Anything else?

>Argh, in the ever increasing scope creep that is this PR I'm pretty sure I just enabled the battery motor compensation code on quadplane. Will need to touch base with @tridge to confirm there aren't any implications of this.

Not by default, but yes, it would be good to know it doesn't cause issues if an user enables it.

WickedShell commented:
Pushed the nitpick fixes already. Your list looks correct to me.

To be fair I was actually expecting some of the motor current limiting stuff to work already on quadplane since I could see the parameters, but I'm glad I wasn't banking on it.

khancyr commented:
checking that the param conversion is working !

WickedShell commented:
@khancyr Great, thanks!

After chatting with tridge the main risk with battery compensation was that quadplanes could use the forward battery rather then the vertical battery, which wouldn't help, and the other concern is if you have a single battery during transition. I've pushed an additional commit that allows AP_Motors to specify what battery index is used (which is primarily useful for multi battery setups), and the default value won't change behavior for anyone.

The other change there is that ` AP_MotorsMulticopter::update_lift_max_from_batt_voltage()` now uses the resting voltage as part of the check rather then the battery voltage. This was checked with Leonard first and it makes more sense to be checking the value we will be using in compensation rather then a different one.

WickedShell commented:
@rmackay9 the main concern was if you could trigger a battery failsafe that actually caused the vehicle to try and do something during the compassmot procedure.

rmackay9 commented:
@WickedShell, I tested attempting to trigger a radio failsafe during the compassmot procedure but wasn't able to.  In particular I did this:

- First ensured that the rc failsafe was working by switching to loiter mode, taking-off (by raising throttle) then turned off the radio.  the RC failsafe correct switched to Land mode (because the vehicle was very near home).
- armed the vehicle but remained landed and tried to start compassmot procedure, it would not start.
- left the vehicle disarmed, started the compassmot procedure and turned off the radio, no failsafe was triggered.  The EKF was very unhappy but I think that's probably because of timing.

WickedShell commented:
@rmackay9 right, that's not the same thing unfortunately. The old code used to explicitly disable the battery failsafes by changing the parameter values. That was removed in this version, however if we need to reintroduce a disable for the cal that's fine, but the modification of parameters was a bad way to do it.

rmackay9 commented:
One small thing is the default BATT_LOW_VOLT is too high.  It shows up as "11" on my IRIS anyway but the default previously was lower (10.4 maybe?).

By the way, i also tried triggering a battery failsafe during compass mot but wasn't able to.

WickedShell commented:
@rmackay9 ah good catch, not sure where I pulled 11V from. Updated. Great to hear that the failsafes aren't an issue while doing the test.

Assuming @OXINARF and others are good, I think this is good.

OXINARF commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -176,6 +176,13 @@ const AP_Param::GroupInfo AP_MotorsMulticopter::var_info[] = {
+    // @Param: BAT_IDX
+    // @DisplayName: Battery compensation index
+    // @Description: Which battery monitor should be used for doing compensation
+    // @Range: 0 1
```
Should this be values instead of range? If you want to go with range, then add the `@Increment: 1`.

OXINARF commented:
Merged, thanks!
